torch/utils/_sympy/functions.py

```diff
@@ -350,45 +586,96 @@ def eval(cls, *args):
         return None
 
 
-class Trunc(sympy.Function):
+# NB: this is inconsistent with math.trunc in Python
+class TruncToFloat(sympy.Function):
+    is_integer = False
+    is_real = True
+
+    @classmethod
+    def eval(cls, number):
+        # assert number.is_integer is not True, number
+        if isinstance(number, sympy.Number):
+            # NB: It is safe to use truncation to integer, which is what
+            # math.trunc does, as Python integers are arbitrary precision and
+            # so we are guaranteed not to lose precision when we do this
+            return sympy.Float(math.trunc(float(number)))
+
+
+class TruncToInt(sympy.Function):
     is_integer = True
 
     @classmethod
     def eval(cls, number):
-        if number.is_integer:
-            return number
-        elif isinstance(number, sympy.Number):
+        # assert number.is_integer is not True, number
+        if number == sympy.oo:
+            return sympy.Integer(sys.maxsize - 1)
+        if number == -sympy.oo:
+            return sympy.Integer(-sys.maxsize - 1)
+        if isinstance(number, sympy.Number):
             return sympy.Integer(math.trunc(float(number)))
 
 
+# This is float -> float.  This is inconsistent with Python builtin round,
+# which returns an integer when there are no digits, but consistent with
+# torch.round, which always returns a float.
+#
+# TODO: As currently written, this is redundant with RoundDecimal, but we
+# leave this open because we have codegen for Round but not
+# RoundDecimal (we do have lowering for RoundDecimal, need to use that...)
 class Round(sympy.Function):
-    is_integer = True
+    is_integer = False
+    is_real = True
 
     @classmethod
     def eval(cls, number):
-        if number.is_integer:
-            return number
-        elif isinstance(number, sympy.Number):
-            return sympy.Integer(round(float(number)))
+        # assert number.is_integer is not True, number
```

Comment:
The not integer tests don't work, see 

```
# It would be nice to have assertions on whether or not inputs is_integer
# However, with bugs like https://github.com/sympy/sympy/issues/26620 sympy
# sometimes inconsistently reports floats an integers.
#   
# What we can assume from sympy is that if something is an int, it
# definitely is is_integer, but if it is a float it may or may not
# be is_integer.  So we are unable to do strong asserts that things
# are NOT integers.
```